lib/registry/src/client.rs

```diff
@@ -0,0 +1,146 @@
+    pub fn with_token(self, token: String) -> Self {
+        Self {
+            token: Some(token),
+            ..self
+        }
+    }
+
+    /// Execute a GraphQL query.
+    async fn execute<Q: GraphQLQuery>(
+        &self,
+        vars: Q::Variables,
+    ) -> Result<graphql_client::Response<Q::ResponseData>, reqwest::Error> {
+        let body = Q::build_query(vars);
+
+        let req = self.client.post(self.endpoint.as_str());
+
+        let req = if let Some(token) = &self.token {
+            req.bearer_auth(token)
+        } else {
+            req
+        };
+        req.json(&body)
+            .send()
+            .await?
+            .error_for_status()?
+            .json::<graphql_client::Response<Q::ResponseData>>()
+            .await
+    }
+
+    /// Execute a GraphQL query, and convert a response with errors to a Rust error.
+    async fn execute_checked<Q: GraphQLQuery>(
+        &self,
+        vars: Q::Variables,
+    ) -> Result<Q::ResponseData, anyhow::Error> {
+        let res = self.execute::<Q>(vars).await?;
+
+        if let Some(data) = res.data {
+            Ok(data)
+        } else {
+            // TODO: Better error forwaring with a custom error type.
+            anyhow::bail!("GraphQL error: {:?}", res.errors);
+        }
+    }
+
+    /// Generate a Deploy token for for the given Deploy app version id.
+    pub async fn generate_deploy_token(
+        &self,
+        app_version_id: String,
+    ) -> Result<String, anyhow::Error> {
+        let vars = graphql::mutations::generate_deploy_token::Variables { app_version_id };
+        let res = self
+            .execute_checked::<graphql::mutations::GenerateDeployToken>(vars)
+            .await?;
+        let token = res
+            .generate_deploy_token
+            .context("Query did not return a token")?
+            .token;
+
+        Ok(token)
+    }
+
+    /// Publish a Deploy app.
+    ///
+    /// Takes a raw, unvalidated deployment config.
+    // TODO: Add a variant of this query that takes a typed DeployV1 config.
+    pub async fn publish_deploy_app_raw(
+        &self,
+        data: PublishDeployAppRawVars,
+    ) -> Result<PublishDeployAppOutput, anyhow::Error> {
+        let vars2 = graphql::mutations::publish_deploy_app::Variables {
+            name: data.name,
+            owner: data.namespace,
+            config: serde_json::to_string(&data.config)?,
+        };
+
+        let version = self
+            .execute_checked::<graphql::mutations::PublishDeployApp>(vars2)
+            .await?
+            .publish_deploy_app
+            .context("Query did not return data")?
+            .deploy_app_version;
+        let app = version.app.context("Query did not return expected data")?;
+
+        Ok(PublishDeployAppOutput {
+            app_id: app.id,
+            app_name: app.name,
+            version_id: version.id,
+            version_name: version.version,
+            owner_name: app.owner.global_name,
```

Comment:
I think this two functions should live outside of the client, as they are just executing mutations but not affecting the global state, while also being domain specific

Reply:
I usually write API clients to be one struct with all the various query methods on the struct instead of separate functions. 

Both designs are valid, but the former is usually more convenient documentation and usability wise, because with all the queries as separate functions things get more boiler-plate heavy and browsing the documentation the available  queries is more cumbersome.

```rust
use registry::Client;

let client = Client::from_env()?;
let user = client.current_user().await?;
let app = client.publish_deploy_app(Vars{ ...}).await?;
let token = client.generate_token(&app.version_id).await?;
```

vs 

```rust
use registry::{
  Client,
  queries,
};

let client = Client::from_env()?;
let user = queries::current_user(&client).await?;
let app = queries::publish_deploy_app(&client, Vars{ ...}).await?;
let token = queries::generate_token(&client, &app.version_id).await?;
```

Reply:
I can be convinced otherwise if you feel strongly about this, but I do usually prefer the first variant.

Reply:
I've reflected a bit on this and I still believe is good to move them out of the client.
Otherwise, the client then starts being responsive on how to do certain queries (but not all) and it's hard to draw the line between when a query should live in the client and when it shouldn't.

Not only that, the second one is more composable frees the concept of queries outside of the client, while the former doesn't

Reply:
@syrusakbary the naming here is a bit difficult, I thought about also using public `queries` and `mutations` submodules, but users of the crate shouldn't have to care about GraphQL details like the queries vs mutations split.

They could also just be root exports, but that's currently suboptimal because of all the other stuff in the crate.

So my compromise was putting them under `api`. 

Happy about suggestions though.